### ISSUE TITLE
ABI checks have problems comparing signed/unsigned numbers

### ISSUE THREAD

**Issue by sdroege** [gtk-rs/gir#572]
See e.g. in GStreamer
```
Constant value mismatch for GST_MESSAGE_ANY
Rust: "4294967295"
C:    "-1"
constants ... 272 passed; 3 failed (compilation errors: 0)
Constant value mismatch for GST_MESSAGE_DEVICE_ADDED
Rust: "2147483649"
C:    "-2147483647"
Constant value mismatch for GST_MESSAGE_DEVICE_REMOVED
Rust: "2147483650"
C:    "-2147483646"
Constant value mismatch for GST_MESSAGE_EXTENDED
Rust: "2147483648"
C:    "-2147483648"
Constant value mismatch for GST_MESSAGE_PROPERTY_NOTIFY
Rust: "2147483651"
C:    "-2147483645"
Constant value mismatch for GST_MESSAGE_REDIRECT
Rust: "2147483654"
C:    "-2147483642"
constants ... 292 passed; 8 failed (compilation errors: 0)
Constant value mismatch for GST_MESSAGE_STREAMS_SELECTED
Rust: "2147483653"
C:    "-2147483643"
Constant value mismatch for GST_MESSAGE_STREAM_COLLECTION
Rust: "2147483652"
C:    "-2147483644"
```

These are all the same value if casting the Rust value to a `c_int` (instead of `c_uint`), or casting the C value the other way around.

CC @EPashkin @tmiasko

**Comment by tmiasko** [comment]
I would consider this to be a bug in g-ir-scanner. It should produce XML with
values in correct range. Otherwise you have no way to determine the underlying
type of enum because it depends on values in enumeration. In this case the
underlying type would be unsigned int when looking at XML, but int when looking
at C (the code actually goes out of the way to prevent changing the underlying
type to unsigned one).

I wonder if all actually works correctly, given that g-ir-repository determines
underlying integer type based on values in enumeration, see giroffsets.c. Though, I
wouldn't be completely surprised it values gets wrapped around before reaching
there.

**Comment by sdroege** [comment]
C enums are always signed or not? Just what integer size is used is implementation defined

**Comment by tmiasko** [comment]
From System V Application Binary Interface for AMD64: 

> C++ and some implementations of C permit enums larger than an int. The underlying
type is bumped to an unsigned int, long int or unsigned long int, in that order.

**Comment by EPashkin** [comment]
Maybe we can add `dont_test` options for object?

**Comment by sdroege** [comment]
These should be tested and the tests would succeed in theory, the binary representation of both numbers is the same.

**Comment by tmiasko** [comment]
1. Rust generator is incorrect in assuming that enumerations always have an int type.
2. g-ir-scanner is incorrect in producing the XML with values outside the range.
3. girepository is actually correctly interpreting the incorrect data produced from XML:

```python
>>> from gi import _gi
>>> from gi.repository import Gst
>>> Gst.MessageType.__info__.get_storage_type()
7
>>> _gi.TypeTag.UINT32
7
```

We could try to stringify actual rust constants and compare them with those
from C (instead of comparing directly with those from XML directly). Then this
particular tests should pass. But, IMHO it would be workaround for above bugs,
that just happen to do correct thing in this particular case. Essentially,
bug 1 happens to cancel out bug 2.

**Comment by tmiasko** [comment]
Is there anything to be done here, or can we close this?

**Comment by EPashkin** [comment]
I still don't see normal solution. IMHO better don't close.

**Comment by sdroege** [comment]
I still think the pragmatic solution here would be to check for value equality, ignoring signedness. That's how these these things are usually handled in the C libraries

**Comment by tmiasko** [comment]
On the other hand, I think that we should use exactly the same underlying
integer type as it is used in C. To do that we need correct enum values, and
any wrapping here is an indication of a bug. This is especially important for
those that don't fit into an integer.

Additionally GstMessageType makes things quite hard here (on both rust generator,
but also on g-ir-scanner), by relaying on C implementation defined behaviour
when casting constants to an integer types where they are non-representable.

I suspect this might be broken in other direction as well, i.e., there are
enums that are represented by unsinged types, whereas Rust uses signed one.
I would to detect those, not hide the issue by wrapping everything.

**Comment by sdroege** [comment]
The question is, what would be the right types? And in C, conversion between signed and unsigned happen implicitly.

I could annotate the flags with different (I.e. the unsigned equivalents) in C if that's more correct. I don't think GI can currently read the correct values itself without annotation anyway (I.e. there already is one).

OTOH internally in GLib all flags types must be compatible with guint, all enums with gint. That's how the GValue machinery is passing them around, so things would fall apart already if this assumption is wrong and all of GLib would fail. IMHO using gint for enums and guint for flags in the Rust bindings would be the right thing in the context of GLib due to that, and if that fails we have problems in GLib to begin with. Also IIRC at least the C# and Python bindings are doing the same (and changing the annotation in GStreamer to the unsigned value might break them).

Tl;dr use guint for flags, gint for enums and if that fails we have to fix GLib anyway. And GLib can't handle enums/flags that are bigger than guint/gint.

**Comment by tmiasko** [comment]
I looked at this in a little bit more detail, mostly from ABI perspective, but
there are also additional constraints from GLib API perspective as you say.

Clang semantic analyzer selects type for an enum, which is also used for code
generation, in following way: if there are negative values, it selects smallest
type from int, long, and long long. Otherwise it selects smallest type from
unsigned int, unsigned long, or unsigned long long. (Things work differently
with packed enums, or with -fshort-enums, but lets ignore that part).

Among gir files I have installed on my system, I didn't find any enum that
would require type larger than int. Which is good news. The only remaining
concern would be about sign. I would presume that sign doesn't matter from ABI
perspective given two's complement representation, so from ABI perspective
we would be already OK.

If we want to go with GLib convention you can just fix GStreamer headers to
use positive values for GMessageType. The GstGLAPI uses something like that
already, i.e., 
```
typedef enum {
  GST_GL_API_NONE = 0,
  ...
  GST_GL_API_ANY = G_MAXUINT32
} GstGLAPI;
```
is it broken in Python?

**Comment by sdroege** [comment]
> Among gir files I have installed on my system, I didn't find any enum that
would require type larger than int. 

It would simply not work with gir because of how GLib's GValue machinery handles them internally. We had to work around this in a few places in GStreamer in the past.

> If we want to go with GLib convention you can just fix GStreamer headers to
use positive values for GstMessageType.

As this is a flags type, it should actually all have unsigned values because of the above. I wonder what would break if that is done.

For GstGL, I don't think that works at all via GIR right now because there's no proper GL.gir :)

> Or they supposed to be signed, but the problem is that they are recognized as flag type?

It's a flags type, which was intentional and a mistake that we can't fix now without breaking API/ABI.


So generally, would you agree that we should "just" assume that flags are guint and enums are gint? Simply because everything else would not be representable by GLib/GIR internals and would fall apart in various ways, and thus be a bug in the C library?

**Comment by sdroege** [comment]
This would basically mean that the testing code on our side in C should print flags as guint and enums as gint, which would indeed solve the problem here.

**Comment by sdroege** [comment]
Basically we would be compiling with `-DABI_CONSTANT_NAME=\(guint\)GST_MESSAGE_ANY` etc for flags, and `-DABI_CONSTANT_NAME=\(gint\)GST_EVENT_EOS` etc for enums to enforce the types expected by GLib.

**Comment by tmiasko** [comment]
I agree that flags should have guint type and other enums have gint type.
The g-ir-scanner should probably emit warnings if this is not the case.

Though, I would still rather validate it, rather then assume it to be true just
to accomdate this singular case, which I still think is a bug in g-ir-scanner
(i.e., it doesn't implement type casts correctly and value in XML is incorrect).

I understand that changing C code might be undesirable, so maybe we just need
a method to skip test for some specific constant / type?

**Comment by sdroege** [comment]
> i.e., it doesn't implement type casts correctly and value in XML is incorrect

What do you mean with this? The type in the XML is simply `GstMessageType` here.

> so maybe we just need a method to skip test for some specific constant / type

The thing is that I'd like to have this specific case also tested, making sure that the values actually match up :) Which they fortunately do after manual checking of the errors, but that seems like something that can be automated.

Maybe what can be checked for is that it works correct when casting to guint/gint, *and* that the size of the uncasted integer type is the same as those two? Signedness in C is rather uninteresting ABI-wise

**Comment by tmiasko** [comment]
> i.e., it doesn't implement type casts correctly and value in XML is incorrect

For example given: `GST_MESSAGE_ANY = (gint) (0xffffffff)`, value of
GST_MESSAGE_ANY in C is -1. If g-ir-scanner were to implement casting correctly
(using the same implementation defined behaviour as in gcc / clang) that is
what it would produce, but it doesn't and instead it writes 4294967295 in gir
XML. This issue is somewhat orthogonal to the whole enum / flag thing, just a
bug in const evaluator.

**Comment by sdroege** [comment]
Ah indeed. If I remember correctly, that cast is there to work around another problem in GI :)
